### PR TITLE
Fix issue 11006 - Make sure the same type is used in pointer ops

### PR DESCRIPTION
**Pull request by LemonBoy**
> If both operands are pointers, and the operator is -, the pointers are subtracted and the result is divided by the size of the type pointed to by the operands. **It is an error if the pointers point to different types**. The type of the result is ptrdiff_t.

### CODE REVIEW

**Comment by BBasile** [comment]
The test is useless. It already works. You should test that it fails now, i.e like the tests in `fail_compilation` that check stderr.

**Comment by denis-sh** [comment]
It's better to also check error message in test.
Also at least DMD source code must be adjusted first (get rid of "Operands must point to the same type" errors) to be able to merge this.

**Comment by LemonBoy** [comment]
Good idea, I've amended the test case.
I've already sent two PRs for druntime and phobos, we just have to wait for them to get merged.

**Comment by WalterWaldron** [comment]
Deprecate first

File: src/expression.d [hunk 1]
```diff
@@ -14159,14 +14159,19 @@ extern (C++) final class MinExp : BinExp
         {
             if (t2.ty == Tpointer)
             {
+                Type p1 = t1.nextOf();
+                Type p2 = t2.nextOf();
+
+                if (!p1.equivalent(p2))
```
Comment (MartinNowak) [inline src/expression.d]: I'd expect this to break a lot of code, maybe rather require types of same size than requiring identity?

File: src/expression.d [hunk 2]
```diff
@@ -14159,14 +14159,19 @@ extern (C++) final class MinExp : BinExp
...
+                    error("Operands must point to the same type (got %s and %s)", p1.toChars(), p2.toChars());
+                    return new ErrorExp();
+                }
+
                 // Need to divide the result by the stride
                 // Replace (ptr - ptr) with (ptr - ptr) / stride
                 d_int64 stride;
 
-                // make sure pointer types are compatible
-                if (Expression ex = typeCombine(this, sc))
```
Comment (MartinNowak) [inline src/expression.d]: Smells like this could do more than just dealing with `void*`.

File: src/expression.d (hunk 1 above)
Comment (LemonBoy) [inline src/expression.d]: Phobos and druntime have been mostly chill as all the errors were due to the use of `void*` and `byte*`. Relying on the size of the types still allows the user to mix and match the types which is not what you want in 99% of the cases (and if you really need to do so you're just explicit cast away) as some types may vary between platforms and one may end up enlarging/shrinking one of the types involved and end up with a not so nice compilation error.
I root for following the specification here.

File: src/expression.d (hunk 2 above)
Comment (LemonBoy) [inline src/expression.d]: It just brings the types down to a common one, I couldn't find a case where something more than that was being done.

**Comment by LemonBoy** [comment]
Should this go trough the deprecation phase too? I think this is a bugfix since we stray away from the specification here.

**Comment by Geod24** [comment]
> Should this go trough the deprecation phase too? I think this is a bugfix since we stray away from the specification here.

Doesn't matter if it's a bug fix or not, if it breaks code and there is a sane depreciation path we have to take it. Remember that a single occurrence in an unused code path in a library will cause downstream projects to fail to compile. That's not something acceptable.

**Comment by yebblies** [comment]
Are pointers with different modifiers handled correctly?  eg `const(void)*` and `void*`

**Comment by LemonBoy** [comment]
Sure, the `equivalent` predicate covers those cases

**Comment by JinShil** [comment]
Revived at #7332

**Comment by LemonBoy** [comment]
You're just grabbing the commits and throwing away all the informations about the author, that's not nice.

**Comment by JinShil** [comment]
> You're just grabbing the commits and throwing away all the informations about the author, that's not nice.

Not at all.  I've posted links in the old PRs to the new PRs, and from the new PRs to the old PRs.  So there is an audit trail that anyone can follow.  In terms of the source code itself, there was no copyright or other information about the author to keep, otherwise I would have kept it.  If you're concerned about the merge being in your name, revive them yourself.

**Pull request closed without merging**